lib/BuildSystem/BuildSystem.cpp

```diff
@@ -1644,33 +1648,27 @@ class ShellCommand : public ExternalCommand {
   bool canSafelyInterrupt = true;
 
   /// The cached signature, once computed -- 0 is used as a sentinel value.
-  std::atomic<uint64_t> cachedSignature{ 0 };
+  std::atomic<CommandSignature> cachedSignature{ };
   
-  virtual uint64_t getSignature() override {
-    uint64_t signature = cachedSignature;
-    if (signature != 0)
+  virtual CommandSignature getSignature() override {
+    CommandSignature signature = cachedSignature;
+    if (!signature.isNull())
       return signature;
-      
-    // FIXME: Use a more appropriate hashing infrastructure.
-    using llvm::hash_combine;
-    llvm::hash_code code = ExternalCommand::getSignature();
-    for (const auto& arg: args) {
-      code = hash_combine(code, arg);
-    }
+
+    auto code = ExternalCommand::getSignature()
+        .combine(!argsForSignature.empty() ? argsForSignature : args);
     for (const auto& entry: env) {
-      code = hash_combine(code, entry.first);
-      code = hash_combine(code, entry.second);
-    }
-    for (const auto& path: depsPaths) {
-      code = hash_combine(code, path);
-    }
-    code = hash_combine(code, int(depsStyle));
-    code = hash_combine(code, int(inheritEnv));
-    code = hash_combine(code, int(canSafelyInterrupt));
-    signature = size_t(code);
-    if (signature == 0) {
-      signature = 1;
-    }
+      code = code.combine(entry.first);
+      code = code.combine(entry.second);
+    }
+    std::for_each(std::begin(depsPaths), std::end(depsPaths), [&code](const auto &path) { code.combine(path); });
+    code = code.combine(int(depsStyle));
+    code = code.combine(int(inheritEnv));
+    code = code.combine(int(canSafelyInterrupt));
+    signature = code;
+//    if (signature.isNull()) {
+//      signature = CommandSignature(1);
+//    }
```

Comment:
I left this here for now because I wanted to bring it up in review: I don't see how the signature could possibly be null at this point, and we don't seem to make a similar check anywhere else in the codebase.

Further, it eliminates one more instance where we initialize a CommandSignature with a raw value of the underlying data type.

Can't we just remove this?

Reply:
OK, I'll leave that in for now.